src/hotspot/share/opto/superword.cpp

```diff
@@ -2737,6 +2718,101 @@ void SuperWord::output() {
+  GrowableArray<SWPointer*> swptrs(_arena, _packset.length(), 0, NULL);
+  for (int i = 0; i < _packset.length(); i++) {
+    Node_List* p = _packset.at(i);
+    assert(p->size() == 1, "all post loop packs should be singleton");
+    Node* n = p->at(0);
+    BasicType bt = velt_basic_type(n);
+    if (n->is_Mem()) {
+      SWPointer* mem_p = new (_arena) SWPointer(n->as_Mem(), this, NULL, false);
+      // For each memory access, we check if the scale (in bytes) in its
+      // address expression is equal to the data size times loop stride.
+      // With this, Only positive scales exist in counting-up loops and
+      // negative scales exist in counting-down loops.
+      if (mem_p->scale_in_bytes() != type2aelembytes(bt) * cl->stride_con()) {
+        return NULL;
+      }
+      swptrs.append(mem_p);
+    }
+    assert(is_java_primitive(bt), "only primitive types are allowed in post loop");
+    stats.record_size(type2aelembytes(bt));
+  }
+
+  // Find the vector data type for generating vector masks. Currently we
+  // don't support post loops with mixed vector data sizes
+  int unique_size = stats.unique_size();
+  BasicType vmask_bt;
+  switch (unique_size) {
+    case 1:  vmask_bt = T_BYTE; break;
+    case 2:  vmask_bt = T_SHORT; break;
+    case 4:  vmask_bt = T_INT; break;
+    case 8:  vmask_bt = T_LONG; break;
+    default: return NULL;
+  }
+
+  // Vector size should be the MaxVectorSize we can use on this machine
+  int vlen = cl->slp_max_unroll();
+  if (unique_size * vlen != MaxVectorSize) {
+    return NULL;
+  }
+
+  // Bail out if target doesn't support mask generator or masked load/store
+  if (!Matcher::match_rule_supported_vector(Op_LoadVectorMasked, vlen, vmask_bt)  ||
+      !Matcher::match_rule_supported_vector(Op_StoreVectorMasked, vlen, vmask_bt) ||
+      !Matcher::match_rule_supported_vector(Op_VectorMaskGen, vlen, vmask_bt)) {
+    return NULL;
+  }
+
+  // Bail out if potential data dependence exists between memory accesses
+  if (SWPointer::has_potential_dependence(swptrs)) {
+    return NULL;
+  }
+
+  // Create vector mask with the post loop trip count. Note there's another
+  // vector drain loop which is cloned from main loop before super-unrolling
+  // so the scalar post loop runs at most vlen-1 trips. Hence, this version
+  // only runs at most 1 iteration after vector mask transformation.
+  Node* trip_cnt;
+  Node* new_incr;
+  if (cl->stride_con() > 0) {
+    trip_cnt = new SubINode(cl->limit(), cl->init_trip());
+    new_incr = new AddINode(cl->phi(), trip_cnt);
+  } else {
+    trip_cnt = new SubINode(cl->init_trip(), cl->limit());
+    new_incr = new SubINode(cl->phi(), trip_cnt);
+  }
+  _igvn.register_new_node_with_optimizer(trip_cnt);
+  _igvn.register_new_node_with_optimizer(new_incr);
+  _igvn.replace_node(cl->incr(), new_incr);
+  Node* offset = new ConvI2LNode(trip_cnt);
+  _igvn.register_new_node_with_optimizer(offset);
+  Node* vmask = VectorMaskGenNode::make(offset, vmask_bt);
```

Comment:
I think `offset` should be `length` here.

Reply:
Renamed, thanks.